runtime/src/iree/builtins/ukernel/arch/arm_64/mmt4d_arm_64_base.c

```diff
@@ -264,3 +264,118 @@ IREE_UK_MMT4D_TILE_FUNC_IMPL_FOR_M0_1_2_4_8(
     iree_uk_mmt4d_tile_s8s8s32_2x8x1_arm_64,
     iree_uk_mmt4d_tile_s8s8s32_4x8x1_arm_64,
     iree_uk_mmt4d_tile_s8s8s32_8x8x1_arm_64)
+
+// This kernel is an adaptation of the kernel
+// `qd8-f32-qc4w-gemm-1x16-minmax-neon-mlal-lane.c` in
+// https://github.com/google/XNNPACK. We borrow the int4 conversion trick
+// described within the method body.
+static inline void iree_uk_mmt4d_tile_s8s4s32_1x16x2_to_4x16x2_arm_64(
+    void* IREE_UK_RESTRICT out_tile, const void* IREE_UK_RESTRICT lhs_panel,
+    const void* IREE_UK_RESTRICT rhs_panel,
+    const iree_uk_mmt4d_params_t* params, int M0) {
+  IREE_UK_ASSERT(M0 >= 1 && M0 <= 4 && iree_uk_is_po2_u32(M0));
+  IREE_UK_ASSERT(!(params->K0 % 2));
+  const iree_uk_int8_t* IREE_UK_RESTRICT lhs_ptr = lhs_panel;
+  const iree_uk_int8_t* IREE_UK_RESTRICT rhs_ptr = rhs_panel;
+
+  iree_uk_int32_t* IREE_UK_RESTRICT out_ptr = out_tile;
+  int32x4_t acc[16];
+  // We start with zero accumulators and add the value of *out_ptr later.
+  // This is required for the int4 trick described later.
+  for (int i = 0; i < 4 * M0; ++i) {
+    acc[i] = vdupq_n_s32(0);
+  }
+
+  const int8x16_t vmask = vmovq_n_s8(0xF0);
+
+  for (int k = 0; k < params->K; ++k) {
+    // Handle 1x16x2.
+    int16_t lhs_0 = *lhs_ptr++;
+    int16_t lhs_1 = *lhs_ptr++;
+    int8x16_t rhs = vld1q_s8(rhs_ptr);
+    rhs_ptr += 16;
+
+    // We unpack int4s into individual int8s. To preserve signedness, int4s are
+    // moved to the upper 4-bits of each byte. This has the effect of
+    // multiplying each int4 by x^4 = 16. To compensate, we divide the
+    // accumulator values by 16 before storing to memory.
+    int8x16_t rhs_low = vshlq_n_s8(rhs, 4);
+    int8x16_t rhs_high = vandq_s8(rhs, vmask);
+
+    int16x8_t rhs_low_0 = vmovl_s8(vget_low_s8(rhs_low));
+    int16x8_t rhs_low_1 = vmovl_s8(vget_high_s8(rhs_low));
+    int16x8_t rhs_high_0 = vmovl_s8(vget_low_s8(rhs_high));
+    int16x8_t rhs_high_1 = vmovl_s8(vget_high_s8(rhs_high));
+
+    acc[0] = vmlal_n_s16(acc[0], vget_low_s16(rhs_low_0), lhs_0);
+    acc[1] = vmlal_n_s16(acc[1], vget_high_s16(rhs_low_0), lhs_0);
+    acc[2] = vmlal_n_s16(acc[2], vget_low_s16(rhs_low_1), lhs_0);
+    acc[3] = vmlal_n_s16(acc[3], vget_high_s16(rhs_low_1), lhs_0);
+
+    acc[0] = vmlal_n_s16(acc[0], vget_low_s16(rhs_high_0), lhs_1);
+    acc[1] = vmlal_n_s16(acc[1], vget_high_s16(rhs_high_0), lhs_1);
+    acc[2] = vmlal_n_s16(acc[2], vget_low_s16(rhs_high_1), lhs_1);
+    acc[3] = vmlal_n_s16(acc[3], vget_high_s16(rhs_high_1), lhs_1);
+
+    if (M0 >= 2) {
+      int16_t lhs_2 = *lhs_ptr++;
+      int16_t lhs_3 = *lhs_ptr++;
+
+      // Handle 2x16x2.
+      acc[4] = vmlal_n_s16(acc[4], vget_low_s16(rhs_low_0), lhs_2);
+      acc[5] = vmlal_n_s16(acc[5], vget_high_s16(rhs_low_0), lhs_2);
```

Comment:
get_high doesnt always implement well on aarch64.  You're hoping there will be a vmla2 that can access the upper part of a register to make this free. vext is slow.  dup may be faster or a mov lane.  But check the code gen to see if this is a problem.
It will work well on armv7 if thats important.

Reply:
Since this kernel will only be used on 5+ year old CPUs, I'm inclined to leave as is and dig deeper once I've written the `+dotprod` and `+i8mm` variants.

Reply:
This kernel is only best for cortex-a75 (aka Pixel 3 big core).
Not the little core of Pixel 3 (a53) or older cpus - a73, a72, a57 or 32 bit cpus like A7.
On the older cpus, its faster to convert the 8 bit to 16 bit and use lanes.

As is, I'm concerned the vget_high_s16 is a bottleneck on aarch64.
It would be okay on aarch32, but you may spill registers.. there are 16 neon registers.
Which is solved by using lanes

Reply:
I moved the `vget_high_s16` and `vget_low_s16` calls outside of the `vmlal_lane_s16` and now creating separate registers of `int16x4_t` so that `vget*` don't get called every time.